### PR TITLE
acquisition: rollover

### PR DESCRIPTION
**Pull request by Garfield-fr**
* Removes the budget change on the organisation.
* Removes all actions in the budget resource.
* Disables actions on the acquisition resources.
* Fixes the title on receipt order lines.
* Closes https://github.com/rero/rero-ils/issues/2929.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

### CODE REVIEW

File: projects/admin/src/app/acquisition/routes/budgets-route.ts [hunk 1]
```diff
@@ -36,11 +34,10 @@ export class BudgetsRoute extends BaseRoute implements RouteInterface {
       matcher: (url: any) => this.routeMatcher(url, this.name),
       children: [
         { path: '', component: RecordSearchPageComponent },
-        { path: 'detail/:pid', component: DetailComponent },
-        { path: 'edit/:pid', component: EditorComponent, canActivate: [CanUpdateGuard] },
```
Comment (jma) [inline projects/admin/src/app/acquisition/routes/budgets-route.ts:40]: Is that means that create or update a budget is not possible anymore?

File: projects/admin/src/app/acquisition/routes/guards/abstract-can-add.guard.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+   * @param _recordService - RecordService
+   * @param _router - Router
+   */
+   protected constructor(
+    protected _recordService: RecordService,
+    protected _router: Router
+  ) {}
+
+  /**
+   * Can add
```
Comment (jma) [inline projects/admin/src/app/acquisition/routes/guards/abstract-can-add.guard.ts:36]: Can you add more details in the comment?

File: projects/admin/src/app/acquisition/routes/guards/can-add-account.guard.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+   * @param _router - Router
+   */
+   constructor(
+    _recordService: RecordService,
+    _router: Router
+  ) {
+    super(_recordService, _router);
+  }
+
+  canActivate(route: ActivatedRouteSnapshot): Observable<boolean> {
```
Comment (jma) [inline projects/admin/src/app/acquisition/routes/guards/can-add-account.guard.ts:40]: Documentation is missing.

File: projects/admin/src/app/acquisition/routes/guards/can-add-order-line.guard.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+   * @param _router - Router
+   */
+   constructor(
+    _recordService: RecordService,
+    _router: Router
+  ) {
+    super(_recordService, _router);
+  }
+
+  canActivate(route: ActivatedRouteSnapshot): Observable<boolean> {
```
Comment (jma) [inline projects/admin/src/app/acquisition/routes/guards/can-add-order-line.guard.ts:40]: Documentation is missing.

File: projects/admin/src/app/acquisition/routes/order-lines-route.ts [hunk 5]
```diff
@@ -90,4 +93,19 @@ export class OrderLinesRoute extends BaseRoute implements RouteInterface {
...
+
+  /**
+   * Pre update record
+   * @param data - Record data
+   * @returns record data
+   */
+  private _preUpdateRecord(data: any): any {
+    const fieldsToRemoved = ['is_current_budget'];
+    fieldsToRemoved.forEach(key => {
+      if (data.hasOwnProperty(key)) {
```
Comment (jma) [inline projects/admin/src/app/acquisition/routes/order-lines-route.ts:105]: `key in data`?

File: projects/admin/src/app/acquisition/routes/receipt-lines-route.ts [hunk 6]
```diff
@@ -76,4 +78,19 @@ export class ReceiptLinesRoute extends BaseRoute implements RouteInterface {
...
+
+  /**
+   * Pre update record
+   * @param data - Record data
+   * @returns record data
+   */
+   private _preUpdateRecord(data: any): any {
+    const fieldsToRemoved = ['is_current_budget'];
+    fieldsToRemoved.forEach(key => {
+      if (data.hasOwnProperty(key)) {
```
Comment (jma) [inline projects/admin/src/app/acquisition/routes/receipt-lines-route.ts:90]: `key in data`?

File: projects/admin/src/app/service/record-permission.service.ts [hunk 7]
```diff
@@ -78,25 +78,28 @@ export class RecordPermissionService {
...
-      });
+
+    if (reasons) {
+      // Links
+      if ('links' in reasons) {
+        const pluralDict = this.plurialLinksMessages();
+        Object.keys(reasons.links).forEach(link => {
+          const message = (link in pluralDict)
+            ? translatePlural.transform(reasons.links[link], pluralDict[link], this._translateService.currentLang)
+            : reasons.links[link][link] + ' ' + link;
```
Comment (jma) [inline projects/admin/src/app/service/record-permission.service.ts:89]: You can use `` instead of string concatenation.

**Review by jma — APPROVED** [review]

File: projects/admin/src/app/acquisition/formly/type/field-ref.type.ts [hunk 8]
```diff
@@ -46,8 +46,17 @@ export class FieldRefTypeComponent extends FieldType implements OnInit {
...
+            const value = this.to.resourceSelect.value;
+            data[keys[0]] = data[keys[0]].filter((element: any) => element[field] === value);
+          }
           keys.forEach((key: string) => {
-            data = data[key];
+            if (data && key in data) {
+              data = data[key];
+            } else {
+              data = undefined;
+            }
```
Comment (zannkukai) [inline projects/admin/src/app/acquisition/formly/type/field-ref.type.ts:59]: cosmetic : you could use a ternary instruction to reduce code
```
data = (data && key in data) ? data[key] : undefined;
```

File: projects/admin/src/app/acquisition/routes/guards/abstract-can-add.guard.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { Router } from '@angular/router';
+import { RecordService } from '@rero/ng-core';
+import { BaseApi } from '@rero/shared';
+import { Observable, of } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+
+export abstract class AbstractCanAddGuard {
```
Comment (zannkukai) [inline projects/admin/src/app/acquisition/routes/guards/abstract-can-add.guard.ts:23]: an abstract class without abstract methods ? sounds strange.

File: projects/admin/src/app/acquisition/routes/guards/can-add-account.guard.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  /**
+   * Allow the loading if the budget pid is present and check
+   * if the record is in current fiscal year
+   * @param route - ActivatedRouteSnapshot
+   * @returns Observable boolean
+   */
+  canActivate(route: ActivatedRouteSnapshot): Observable<boolean> {
+    const budgetPid = route.queryParams.budget;
+    if (!budgetPid) {
+      this._router.navigate(['/errors/404'], { skipLocationChange: true });
```
Comment (zannkukai) [inline projects/admin/src/app/acquisition/routes/guards/can-add-account.guard.ts:49]: is 404 is the right code ? 400 (bad request) or 401 (forbidden) could be better ?

File: projects/admin/src/app/acquisition/routes/guards/can-add-order-line.guard.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  /**
+   * Allow the loading if the order pid is present and check
+   * if the record is in current fiscal year
+   * @param route - ActivatedRouteSnapshot
+   * @returns Observable boolean
+   */
+  canActivate(route: ActivatedRouteSnapshot): Observable<boolean> {
+    const orderPid = route.queryParams.order;
+    if (!orderPid) {
+      this._router.navigate(['/errors/404'], { skipLocationChange: true });
```
Comment (zannkukai) [inline projects/admin/src/app/acquisition/routes/guards/can-add-order-line.guard.ts:49]: same here

File: projects/admin/src/app/acquisition/routes/guards/can-budget-active.guard.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    const pid = route.params.pid;
+    if (!type || !pid) {
+      return of(false);
+    }
+    return this._recordService.getRecord(type, pid, 0, BaseApi.reroJsonheaders).pipe(
+      map(record => {
+        if ('metadata' in record) {
+          return record.metadata;
+        }
+        return record;
```
Comment (zannkukai) [inline projects/admin/src/app/acquisition/routes/guards/can-budget-active.guard.ts:55]: single instruction using ternary : `map(record => ('metadata' in record) ? record.metadata : record)`

File: projects/admin/src/app/acquisition/routes/guards/can-budget-active.guard.ts [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+import { ActivatedRouteSnapshot, CanActivate, Router } from '@angular/router';
+import { RecordService } from '@rero/ng-core';
+import { BaseApi } from '@rero/shared';
+import { Observable, of } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CanBudgetActiveGuard implements CanActivate {
```
Comment (zannkukai) [inline projects/admin/src/app/acquisition/routes/guards/can-budget-active.guard.ts:27]: Reflexion : Don't know if the class name sounds good ; maybe "isBudgetActiveGuard" ? I suggest, you choose ;-)

File: projects/admin/src/app/acquisition/routes/receipt-lines-route.ts [hunk 14]
```diff
@@ -76,4 +78,19 @@ export class ReceiptLinesRoute extends BaseRoute implements RouteInterface {
...
+   * @returns record data
+   */
+   private _preUpdateRecord(data: any): any {
+    const fieldsToRemoved = ['is_current_budget'];
+    fieldsToRemoved.forEach(key => {
+      if (key in data) {
+        delete data[key];
+      }
+    });
+    return data;
```
Comment (zannkukai) [inline projects/admin/src/app/acquisition/routes/receipt-lines-route.ts:94]: Suggestion : As this method seems repetitive for some routes, could be useful to place it into `BaseRoute` with `fieldsToRemoved` as required parameter.

File: projects/admin/src/app/acquisition/routes/receipt-lines-route.ts [hunk 15]
```diff
@@ -76,4 +78,13 @@ export class ReceiptLinesRoute extends BaseRoute implements RouteInterface {
...
   }
+
+  /**
+   * Pre update record
+   * @param data - Record data
+   * @returns record data
+   */
+   private _preUpdateRecord(data: any): any {
+    return this.fieldsToRemoved(data, ['is_current_budget']);
+  }
```
Comment (zannkukai) [inline projects/admin/src/app/acquisition/routes/receipt-lines-route.ts:89]: Why not just use `this.fieldsToRemoved(data, ['is_current_budget']);` into the `preUpdateRecord` (Line 50) (same remark for other routes) ?

**Review by zannkukai — APPROVED** [review]

File: projects/admin/src/app/acquisition/routes/receipt-lines-route.ts (hunk 15 above)
Comment (Garfield-fr) [reply projects/admin/src/app/acquisition/routes/receipt-lines-route.ts:89]: I could. I just thought I'd put the same name of the function I'm calling in the route. There are other places where I find the name is not OK, but we can always improve.

**Pull request merged**